R/generate_dust.R

```diff
@@ -0,0 +1,222 @@
+    x$depends$functions), FALSE, FALSE))
+  support <- NULL
+  if ("odin_sum" %in% used) {
+    ranks <- sort(unique(viapply(dat$data$elements, "[[", "rank")))
+    ranks <- ranks[ranks > 0]
+    if (length(ranks) > 0L) {
+      support <- c(support, lapply(ranks, generate_dust_support_sum))
+    }
+  }
+
+  list(class = class, create = create, support = support)
+}
+
+
+generate_dust_meta <- function() {
+  list(rng = "rng")
+}
+
+
+generate_dust_core_class <- function(eqs, dat, rewrite) {
+  struct <- generate_dust_core_struct(dat)
+  ctor <- generate_dust_core_ctor(dat)
+  size <- generate_dust_core_size(dat, rewrite)
+  initial <- generate_dust_core_initial(dat, rewrite)
+  update <- generate_dust_core_update(eqs, dat, rewrite)
+
+  ret <- collector()
+  ret$add("class %s {", dat$config$base)
+  ret$add("public:")
+  ret$add(paste0("  ", struct))
+  ret$add(paste0("  ", ctor))
+  ret$add(paste0("  ", size))
+  ret$add(paste0("  ", initial))
+  ret$add(paste0("  ", update))
+  ret$add("private:")
+  ret$add("  init_t %s;", dat$meta$internal)
+  ret$add("};")
+
+  ret$get()
+}
+
+
+generate_dust_core_struct <- function(dat) {
+  struct_element <- function(x) {
+    type <- x$storage_type
+    is_ptr <- x$rank > 0L || type == "void"
+    if (is_ptr) {
+      sprintf("std::vector<%s> %s;", type, x$name)
+    } else {
+      sprintf("%s %s;", type, x$name)
+    }
+  }
+  i <- vcapply(dat$data$elements, "[[", "location") == "internal"
+  els <- vcapply(unname(dat$data$elements[i]), struct_element)
+
+  c("struct init_t {",
+    paste0("  ", els),
+    "};")
+}
+
+
+generate_dust_core_ctor <- function(dat) {
+  c(sprintf("%s(const init_t& data): %s(data) {",
+            dat$config$base, dat$meta$internal),
+    "}")
+}
+
+
+generate_dust_core_size <- function(dat, rewrite) {
+  body <- sprintf("return %s;", rewrite(dat$data$variable$length))
+  cpp_function("size_t", "size", NULL, body)
+}
+
+
+generate_dust_core_initial <- function(dat, rewrite) {
+  set_initial <- function(el) {
+    data_info <- dat$data$elements[[el$name]]
+    if (data_info$rank == 0L) {
+      lhs <- sprintf("%s[%s]", dat$meta$state, rewrite(el$offset))
+      sprintf("%s = %s.%s;", lhs, dat$meta$internal, el$initial)
+    } else {
+      src <- rewrite(el$initial)
+      sprintf(
+        "std::copy(%s.begin(), %s.end(), %s.begin() + %s);",
+        src, src, dat$meta$state, rewrite(el$offset))
+    }
+  }
+
+  if (length(dat$components$initial$equations)) {
+    subs <- lapply(dat$data$variable$contents, function(x) rewrite(x$initial))
+    eqs_initial <- dat$equations[dat$components$initial$equations]
+    eqs_initial <- lapply(odin:::ir_substitute(eqs_initial, subs),
+                          generate_dust_equation, dat, rewrite)
+  } else {
+    eqs_initial <- NULL
+  }
+
+  initial <- dust_flatten_eqs(lapply(dat$data$variable$contents, set_initial))
+
+  args <- c("size_t" = dat$meta$time)
+  body <- c(sprintf("std::vector<double> %s(%s);",
```

Comment:
yes, I think doing this in one concerted push would be sensible. I have not written code that uses `float` rather than `double` perhaps ever! Seems like that's something that we'll hit with the GPU work?

Reply:
I did a short course on C++ optimisation and the instructor said `float` vs `double` makes almost no performance difference with modern CPUs, so you should always go for the higher precision. The only time I've used it is to save 50% memory with large genomic datasets.

However, the CUDA best practises guide says to use single precision over double precision, if that precision loss is acceptable: https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#overall-performance-optimization-strategies
I believe it's mostly a memory bandwidth issue, as you can fit twice as many thread's data on the bus when reading/writing from global memory, but I bet there are also more complex issues (in the the math library there are overloads for half, single and double precision arithmetic which speed things up).

Reply:
Logged as https://github.com/mrc-ide/odin.dust/issues/4 which depends on https://github.com/mrc-ide/dust/issues/27